go-controller/pkg/ovn/controller/services/load_balancer.go

```diff
@@ -0,0 +1,456 @@
+func makeLBName(service *v1.Service, proto v1.Protocol, scope string) string {
+	return fmt.Sprintf("Service_%s/%s_%s_%s",
+		service.Namespace, service.Name,
+		proto, scope,
+	)
+}
+
+// buildClusterLBs takes a list of lbConfigs and aggregates them
+// in to one ovn LB per protocol.
+//
+// It takes a list of (proto:[vips]:port -> [endpoints]) configs and re-aggregates
+// them to a list of (proto:[vip:port -> [endpoint:port]])
+// This load balancer is attached to all node switches. In shared-GW mode, it is also on all routers
+func buildClusterLBs(service *v1.Service, configs []lbConfig, nodeInfos []nodeInfo) []ovnlb.LB {
+	nodeSwitches := make([]string, 0, len(nodeInfos))
+	nodeRouters := make([]string, 0, len(nodeInfos))
+	for _, node := range nodeInfos {
+		nodeSwitches = append(nodeSwitches, node.switchName)
+		if globalconfig.Gateway.Mode == globalconfig.GatewayModeShared && node.gatewayRouterName != "" {
+			nodeRouters = append(nodeRouters, node.gatewayRouterName)
+		}
+	}
+
+	cbp := configsByProto(configs)
+
+	out := []ovnlb.LB{}
+	for _, proto := range protos {
+		cfgs, ok := cbp[proto]
+		if !ok {
+			continue
+		}
+		lb := ovnlb.LB{
+			Name:        makeLBName(service, proto, "cluster"),
+			Protocol:    string(proto),
+			ExternalIDs: util.ExternalIDsForObject(service),
+			Opts:        lbOpts(service),
+
+			Switches: nodeSwitches,
+			Routers:  nodeRouters,
+		}
+
+		for _, config := range cfgs {
+			if config.routerLocalMode {
+				panic("coding error: local endpoints only for cluster-wide lbConfig")
+			}
+
+			v4targets := make([]ovnlb.Addr, 0, len(config.eps.V4IPs))
+			for _, tgt := range config.eps.V4IPs {
+				v4targets = append(v4targets, ovnlb.Addr{
+					IP:   tgt,
+					Port: config.eps.Port,
+				})
+			}
+
+			v6targets := make([]ovnlb.Addr, 0, len(config.eps.V6IPs))
+			for _, tgt := range config.eps.V6IPs {
+				v6targets = append(v6targets, ovnlb.Addr{
+					IP:   tgt,
+					Port: config.eps.Port,
+				})
+			}
+
+			rules := make([]ovnlb.LBRule, 0, len(config.vips))
+			for _, vip := range config.vips {
+				if vip == "node" {
+					panic("coding error: node IP for cluster LB")
+				}
+				targets := v4targets
+				if strings.Contains(vip, ":") {
+					targets = v6targets
+				}
+
+				rules = append(rules, ovnlb.LBRule{
+					Source: ovnlb.Addr{
+						IP:   vip,
+						Port: config.inport,
+					},
+					Targets: targets,
+				})
+			}
+			lb.Rules = append(lb.Rules, rules...)
+		}
+
+		out = append(out, lb)
+	}
+	return out
+}
+
+// buildPerNodeLBs takes a list of lbConfigs and expands them to one LB per protocol per node
+// This works differently based on whether or not we're in shared or local gateway mode.
+//
+// For local gateway, per-node lbs are created for:
+// - nodePort services, attached to each node's switch, vips are node's physical IPs
+//
+// For shared gateway, per-node lbs are created for
+// - clusterip services with host-network services are attached to each node's gateway router + switch
```

Comment:
nit: host-network endpoints

Reply:
totally not a nit; I tried for these comments to be meaningful.

fixed.